django/librarian/views.py

```diff
@@ -470,8 +471,18 @@ def upload(request, data_source_id):
     bind_contextvars(feature="librarian")
 
     for file in request.FILES.getlist("file"):
-        file_obj = SavedFile.objects.create(content_type=file.content_type)
-        file_obj.file.save(file.name, file)
+        # Check if the file is already stored on the server
+        file_hash = generate_hash(file.read())
+        file_exists = SavedFile.objects.filter(sha256_hash=file_hash).exists()
+        if file_exists:
+            file_obj = SavedFile.objects.filter(sha256_hash=file_hash).first()
+            logger.info(
+                f"Found existing SavedFile for {file.name}", saved_file_id=file_obj.id
+            )
+        else:
+            file_obj = SavedFile.objects.create(content_type=file.content_type)
+            file_obj.file.save(file.name, file)
+            file_obj.generate_hash()
```

Comment:
The basic librarian file upload needs to also be replaced with JS version, like in the Chat upload above, but I will do that in another PR.